File: padroesEstruturais/padrao-de-projeto-composite/java/src/test/java/ComputadorBuilderTest.java
import composite.Componentes.Fonte;
import composite.Componentes.PlacaMae;
import composite.Componentes.Processador;
import composite.Computador;
import composite.Pecas.Parafuso;
import composite.builders.ComputadorBuilder;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class ComputadorBuilderTest {

    @Test
    public void deveRetornarExcecaoParaComputadorSemPecas() {
        try {
            ComputadorBuilder computadorBuilder = new ComputadorBuilder();
            computadorBuilder.build();
            fail();
        } catch (IllegalArgumentException exception) {
            assertEquals("Computador sem peças", exception.getMessage());
        }
    }

    @Test
    public void deveRetornarExcecaoParaComputadorIncompletoSemPlacaMae() {
        try {
            ComputadorBuilder computadorBuilder = new ComputadorBuilder().setComponents(new Processador()).setComponents(new Fonte());
            computadorBuilder.build();
            fail();
        } catch (IllegalArgumentException exception) {
            assertEquals("Computador incompleto", exception.getMessage());
        }
    }

    @Test
    public void deveRetornarExcecaoParaComputadorIncompletoSemProcessador() {
        try {
            ComputadorBuilder computadorBuilder = new ComputadorBuilder().setComponents(new PlacaMae()).setComponents(new Fonte());
            computadorBuilder.build();
            fail();
        } catch (IllegalArgumentException exception) {
            assertEquals("Computador incompleto", exception.getMessage());
        }
    }

    @Test
    public void deveRetornarExcecaoParaComputadorIncompletoSemFonte() {
        try {
            ComputadorBuilder computadorBuilder = new ComputadorBuilder().setComponents(new PlacaMae()).setComponents(new Fonte());
            computadorBuilder.build();
            fail();
        } catch (IllegalArgumentException exception) {
            assertEquals("Computador incompleto", exception.getMessage());
        }
    }

    @Test
    public void deveRetornarComputadorComponentes() {
        ComputadorBuilder computadorBuilder = new ComputadorBuilder();
        computadorBuilder
            .setComponents(
                    new PlacaMae().setDescricao("Placa mãe").setValor(1000f).setPecas(new Parafuso().setDescricacao("parafuso"))
            )
            .setComponents(
                    new Processador().setDescricao("Processador").setValor(500f).setPecas(new Parafuso().setDescricacao("parafuso"))
            )
            .setComponents(
                    new Fonte().setDescricao("Fonte").setValor(200f).setPecas(new Parafuso().setDescricacao("parafuso"))
            );

        Computador computador = computadorBuilder.build();
        assertEquals("Descrição: \n" +
            "Placa mãe / valor: 1000.0\n" +
            "Peças: \n" +
            "parafuso\n" +
            "\n" +
            "Processador / valor: 500.0\n" +
            "Peças: \n" +
            "parafuso\n" +
            "\n" +
            "Fonte / valor: 200.0\n" +
            "Peças: \n" +
            "parafuso\n\n",
            computador.getDescricao()
        );
    }
}

File: padroesEstruturais/padrao-de-projeto-composite/php/tests/unit/ComputadorBuilderTest.php
<?php

namespace unit;
use Composite\Php\Builders\ComputadorBuilder;
use Composite\Php\Componentes\Fonte;
use Composite\Php\Componentes\PlacaMae;
use Composite\Php\Componentes\Processador;
use Composite\Php\Pecas\Parafuso;
use PHPUnit\Framework\TestCase;


class computadorBuilderTest extends TestCase
{

    public function testDeveRetornarExcecaoParaComputadorSemPecas(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage("Computador sem peças");
        $computadorBuilder = new ComputadorBuilder();
        $computadorBuilder->build();
    }

    public function testDeveRetornarExcecaoParaComputadorIncompletoSemPlacaMae(): void
    {
        try {
            $computadorBuilder = (new ComputadorBuilder())
                ->setComponentes(new Processador())
                ->setComponentes(new Fonte());

            $computadorBuilder->build();
            fail();
        } catch (\InvalidArgumentException $exception) {
            $this->assertEquals("Computador incompleto", $exception->getMessage());
        }
    }

    public function testDeveRetornarExcecaoParaComputadorIncompletoSemProcessador(): void
    {
        try {
            $computadorBuilder = (new ComputadorBuilder())
                ->setComponentes(new PlacaMae())
                ->setComponentes(new Fonte());

            $computadorBuilder->build();
            fail();
        } catch (\InvalidArgumentException $exception) {
            $this->assertEquals("Computador incompleto", $exception->getMessage());
        }
    }

    public function testDeveRetornarExcecaoParaComputadorIncompletoSemFonte(): void
    {
        try {
            $computadorBuilder = (new ComputadorBuilder())
                ->setComponentes(new PlacaMae())
                ->setComponentes(new Processador());

            $computadorBuilder->build();
            fail();
        } catch (\InvalidArgumentException $exception) {
            $this->assertEquals("Computador incompleto", $exception->getMessage());
        }
    }

    public function testDeveRetornarComputadorComComponentes(): void
    {
        $computadorBuilder = new ComputadorBuilder();
        $computadorBuilder
            ->setComponentes(
                (new PlacaMae())->setDescricao("Placa mãe")->setValor(1000)->setPecas((new Parafuso())->setDescricao("parafuso"))
            )
            ->setComponentes(
                (new Processador())->setDescricao("Processador")->setValor(500)->setPecas((new Parafuso())->setDescricao("parafuso"))
            )
            ->setComponentes(
                (new Fonte())->setDescricao("Fonte")->setValor(200)->setPecas((new Parafuso())->setDescricao("parafuso"))
            );

        $computador = $computadorBuilder->build();
        $this->assertEquals("Descrição: \n" .
            "Placa mãe / valor: 1000\n" .
            "Peças: \n" .
            "parafuso\n" .
            "\n" .
            "Processador / valor: 500\n" .
            "Peças: \n" .
            "parafuso\n" .
            "\n" .
            "Fonte / valor: 200\n" .
            "Peças: \n" .
            "parafuso\n\n",
            $computador->getDescricao()
        );
    }
}